addons/website_sale_loyalty/controllers/main.py

```diff
@@ -103,7 +103,8 @@ def claim_reward(self, reward_id, code=None, **post):
                 coupon = coupon_
                 if code == coupon.code and (
                     (program_sudo.trigger == 'with_code' and program_sudo.program_type != 'promo_code')
-                    or (program_sudo.trigger == 'auto' and program_sudo.applies_on == 'future')
+                    or (program_sudo.trigger == 'auto' and program_sudo.applies_on == 'future'
+                        and program_sudo.program_type not in ('ewallet', 'gift_card', 'loyalty'))
```

Comment:
Is the `gift_card` also missing in master then?
Why is this necessary for this bugfix?

Reply:
What do you mean?

Reply:
1. This change is already present in master, but without the `gift_card`
2. I was wondering how this is related to this bugfix, as it's not exactly the same problem, this is mostly curiosity ^^

Reply:
1. The change in master was actually for a different thing and by accident fix half of the problem.
2. Indeed I checked again and the gift card one is different as with this fix I would break gift card totally. 
Will make take care of that one in different fix both for 16.3 and master.